### ISSUE TITLE
Feature Request: Configuration option for marking connections as "broken" after a JedisDataException

### ISSUE THREAD

**Issue by terrysuereth-glu** [xetorthio/jedis#1087]
Background information: In our use-case, redis redundancy is accomplished by running a single "master" node and multiple read-only replica nodes.  When a failure occurs on the master, it is demoted from master status (becoming read-only) and one of the replicas is then promoted (gaining writability) to become the new master.  This master switch is flipped by using CNAME name records, i.e., clients connecting to the redis cluster at "redis.host.name" will subsequently be pointed to whichever node is the current master.  (I do not particularly like this setup, but the infrastructure is out of my and my colleagues' hands.)

The problem we encounter in this case is when an issue occurs on the master node that _does not sever active connections._  The master node becomes demoted and read-only, but any clients which were previously connected to it are still connected after the demotion.  (Or, alternatively, a race condition may occur where the client is able to reconnect to this redis node before the demotion event is effective.)  Then when a client attempts a write operation on this node, it is knocked back with a JedisDataException due to the redis node being read-only.

If this was a JedisConnectionException, then the connection would automatically be marked as broken by the catch block in Connection.sendCommand, and the connection resource would be cleaned up once the Jedis object is scoped out.  However, since a JedisDataException won't trigger this broken flag, our client's connection will return to the pool and be re-used - continuing to trigger the same error/exception - indefinitely.

In other words, in our use-case, a **JedisDataException is a _de facto_ connection error.**  We must react to it by closing the connection and creating a new one, which should hit the new master redis node.

The workaround in our use-case is to explicitly catch both JedisConnectionException and JedisDataException, and to use **releaseBrokenResource (now deprecated)** in both cases to mark the connection as broken.  This also makes us unable to use a try-with-resources pattern to initialize and auto-close a Jedis connection, since in the error case, the connection would be closed twice.

I realize, with lament, that this use-case is highly specific and the behavior I want is not generally ideal.  So my feature request is to add a configuration option to JedisPool to enable a "mark connection as broken after a JedisDataException" behavior.  E.g. if a Connection has "tolerateDataExceptions" set to false, a JedisDataException in Connection.sendCommand will cause the Connection to be marked as broken.

**Comment by HeartSaVioR** [comment]
@terrysuereth-glu 
Thanks for reporting!

I skimmed your document, and feel it is some kind of specific use-case, so I wonder we can handle it as general option.

But, I also feel that we should have a interface to mark connection as broken, since we can't discard Jedis instance from the pool manually when releaseBrokenResource is removed.

Another improvement idea is setting exception handler for validating connection per each result of operation. It would be not easy to add, so we may have a enough time to consider.

@marcosnils I'd like to hear your thought. Thanks!

**Comment by marcosnils** [comment]
@terrysuereth-glu @HeartSaVioR the idea of adding a configuration parameter to JedisPool doesn't seem like the best solution because even though when after all connections are re-established against the corresponding master node, you might still get `JedisDataExceptions` for other incorrect operations  which would trigger a new reconnection unnecessarily. 

Now, regarding to @terrysuereth-glu specific use case I guess you don't have too many alternatives because even if we add a way to set the `broken` state manually, then you still won't be able to use try with resources as the `Jedis` instance will be outside the `try` scope.

IE: 
```java
JedisPool jp = new JedisPool("localhost", 6370);
try (Jedis j = jp.getResource()) {
    j.set("foo", "bar");
} catch (JedisDataException jde) {
    if (isReadOnlyError(jde)) { // Checks for readonly error
        j.setBroken(true) // You don't have access to the j variable here
    }

}
```

So you still won't be able to use the automatic `close` method. 

As @HeartSaVioR said, seems like you're running a very particular case where you need `Jedis` specifically to return a `JedisConnectionException` when `Read-only` error is returned. What I'd do if I were in your situation is sub-classing `Jedis` or `BinaryJedis` (depending on what you need)  override the `sendCommand` method and add my custom error handling there (in your case, throwing a JedisConnectionException upon "readonly" error). Then you can just create your own JedisPool of this sub-class and use try-with-resources normally. 

In reference to manually flagging an instance as `broken`, it's an idea that I'm not sure that's the best. Because if we handle the `broken` state, then internally we might do some custom logic knowing beforehand what `broken` implies. If we allow to user to happily modify this value, then we might see unknown behaviors because of this.

**Comment by HeartSaVioR** [comment]
@marcosnils 
@terrysuereth-glu has a particular case, but we also know that Redis sends ```ERR max number of clients reached```, which we should mark it as broken. #1036 
I think it is a common case.

**Comment by HeartSaVioR** [comment]
Can't we access j to catch cause? That's really bad.

Maybe off-topic, writing a program with Scala (though I'm totally beginner to Scala) and back to Java I'm really annoyed. I wonder upgrading to Java 8 can get rid of it.

**Comment by marcosnils** [comment]
@HeartSaVioR you're correct. Maybe a simple approach would be to include the `Jedis` instance inside the exception and as I said before, to allow to set the Client in a broken state. 

That way, something like this could be done: 

```java
JedisPool jp = new JedisPool("localhost", 6370);
try (Jedis j = jp.getResource()) {
    j.set("foo", "bar");
} catch (JedisDataException jde) {
    if (isReadOnlyError(jde)) { // Checks for readonly error
        jde.getJedisInstance().setBroken(true)
    }
}
```

I prefer to delegate to the user which exceptions he considers that should flag the connections as it has been broken instead of doing that work ourselves. I like to idea to keep our exception handling as simple as possible with is: 
- Everything that redis returns as an error will be raised as a JedisDataException
- Any connection errors will be handled as JedisConnectionExceptions

**Comment by HeartSaVioR** [comment]
@marcosnils Agreed. Let's not include parsing message to Jedis itself. We can document it instead.

**Comment by terrysuereth-glu** [comment]
Thanks for the comments and discussion.

> ... you might still get JedisDataExceptions for other incorrect operations which would trigger a new reconnection unnecessarily.

This is a good point.  There is certainly a theoretical risk of some other condition that triggers a JedisDataException leading to an infinite loop of disconnects and reconnects (to our master node), when that might not necessarily be correct.

That "setBroken()" proposal sounds good to me, it accommodates our corner case - and potentially others like it - while allowing our code to stay outside of Jedis's understanding of connection opening and closing.

**Comment by marcosnils** [comment]
@HeartSaVioR @terrysuereth-glu doing some tests I just realized that setting the `broken` state manually is not enough because the `try-with-resources` block executes the `close` method on the object __before__ any `catch` or `finally` block. So in our case even if I set the `broken` state in the `catch` block, the `close` method will be executed before and the connection will be returned to the pool normally. 

I guess the only option here is to undeprecate the `returnXXXResource` in JedisPool and allow users to execute those actions manually (pretty much as @terrysuereth-glu is doing today). We should however document that those methods should be used only for custom situations and not as a regular basis.

Thoughts?

**Comment by HeartSaVioR** [comment]
@marcosnils @terrysuereth-glu 
If close() method is called before any catch block, we can't even call returnXXXResource() since it is already returned to pool, and we're returning object to pool twice.
So, it requires users to not use ```try-with-resources```, and it means we should also un-deprecate returnResource() too.

I think if we find a way to live with ```try-with-resources```, it would be better.

IMO, ```custom exception handler for validating connection``` still make sense and compatible with try-with-resources.
It is only required for pooled instance, so we can pass handler into JedisFactory, and set handler to Jedis whenever we create instance from factory.

**Comment by marcosnils** [comment]
@HeartSaVioR I was suggesting keeping the traditional `try catch block` and use the `reutrnXXXResource()` as @terrysuereth-glu is using nowadays. 

I also agree that trying to support `try-with-resources` is the best way to go, but I can't imagine how to implement a `custom exception handler` to handle this situations?

Can you share a snippet about what you have in mind?

**Comment by HeartSaVioR** [comment]
@marcosnils I'll think more and sketch when I'm sure it makes sense.

**Comment by HeartSaVioR** [comment]
@terrysuereth-glu @marcosnils Please see #1101 and comment. Thanks.
